app/services/course/duplication/course_duplication_service.rb

```diff
@@ -0,0 +1,64 @@
+# frozen_string_literal: true
+
+class Course::Duplication::CourseDuplicationService < Course::Duplication::BaseService
+  class << self
+    # Constructor for the course duplication service.
+    #
+    # @param [Course] current_course The course to duplicate.
+    # @param [Hash] options The options to be sent to the Duplicator object.
+    # @option options [User] :current_user (+User.system+) The user triggering the duplication.
+    # @option options [String] :new_title ('Duplicated') The title for the duplicated course.
+    # @option options [DateTime] :new_start_date Start date for the duplicated course.
+    # @param [Array] all_objects All the objects in the course.
+    # @param [Array] selected_objects The objects to duplicate.
+    # @return [Course] The duplicated course
+    def duplicate_course(current_course, options = {}, all_objects = [], selected_objects = [])
+      # TODO: To remove survey from excluded objects after survey duplication is implemented
+      excluded_objects = all_objects - selected_objects | current_course.surveys
+      options[:excluded_objects] = excluded_objects
+      options[:current_course] = current_course
+      options[:time_shift] =
+        if options[:new_start_date]
+          Time.zone.parse(options[:new_start_date]) - current_course.start_at
+        else
+          0
+        end
+      options.reverse_merge!(DEFAULT_COURSE_DUPLICATION_OPTIONS)
+      service = new(:course, options)
```

Comment:
I think it would be cleaner to merge `:course` into `options`, and `options` already have a `mode: :course`.